File: csharp/factory_method.cs
using System;

abstract class Animal {
    public abstract string Speak();
}

class Dog : Animal {
    public override string Speak() {
        return "Woof!";
    }
}

class Cat : Animal {
    public override string Speak() {
        return "Meow!";
    }
}

class AnimalFactory {
    public static Animal CreateAnimal(string type) {
        if (type == "dog") return new Dog();
        else if (type == "cat") return new Cat();
        else return null;
    }
}

class Program {
    static void Main(string[] args) {
        Animal animal = AnimalFactory.CreateAnimal("cat");
        if (animal != null) Console.WriteLine(animal.Speak());
    }
}

File: cpp/factory_method.cpp
#include <iostream>
#include <memory>
#include <string>

class Animal {
public:
    virtual std::string speak() = 0;
    virtual ~Animal() {}
};

class Dog : public Animal {
public:
    std::string speak() override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    std::string speak() override {
        return "Meow!";
    }
};

class AnimalFactory {
public:
    static std::unique_ptr<Animal> createAnimal(const std::string& type) {
        if (type == "dog") return std::make_unique<Dog>();
        else if (type == "cat") return std::make_unique<Cat>();
        else return nullptr;
    }
};

int main() {
    auto animal = AnimalFactory::createAnimal("dog");
    if (animal) std::cout << animal->speak() << std::endl;
}
